# Pandas Assignment

**Name:** Varshini Priya  
**Roll No:** 238T1A0555

## Section A – Output Prediction

### Q1. Array Operations and Vectorization

**Question 1:** Predict the output of both statements.

**Question 2:** Identify one similarity and one difference between the two data structures.

In [ ]:
import numpy as np
import pandas as pd

arr = np.array([10, 20, 30])
s = pd.Series([10, 20, 30])

print(arr * 2)
print(s * 2)

### Q1 Answer

1. The NumPy array output is [20 40 60].
   The Pandas Series output is:
   0    20
   1    40
   2    60
   dtype: int64

2. Similarity: Both support element-wise multiplication.

   Difference: A Pandas Series has an index, while a NumPy array does not have a Pandas index.

**Pandas operation:** Multiplying a Pandas Series by 2 performs element-wise multiplication. Each value in the Series is multiplied by 2, while the original index is preserved.

### Q2. Boolean Filtering

**Question 1:** Predict the rows displayed.

**Question 2:** How many rows and columns will the resulting DataFrame contain?

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Student': ['A', 'B', 'C', 'D'],
    'Marks': [45, 80, 65, 90]
})

result = df[df['Marks'] >= 65]

print(result[['Student', 'Marks']])

### Answer

**1. Predict the rows displayed:**

```text
  Student  Marks
1       B     80
2       C     65
3       D     90
```

**2. How many rows and columns will the resulting DataFrame contain?**

- **Rows:** 3
- **Columns:** 2

**Pandas operation:** The condition `df['Marks'] >= 65` filters the DataFrame and selects only the rows where Marks are 65 or above.

### Q3. Descriptive Statistics

**Question 1:** Predict the mean, maximum, and median of the Marks column.

**Question 2:** Explain the difference between median and mean in this context.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Marks': [10, 20, 30, 40, 50]
})

print(df['Marks'].mean())
print(df['Marks'].max())
print(df['Marks'].median())

### Answer

**1. Predicted Output:**

```text
30.0
50
30.0
```

- **Mean:** 30.0
- **Maximum:** 50
- **Median:** 30.0

**2. Difference between mean and median:**  
The mean is the average of all marks, while the median is the middle value when the marks are arranged in order.

**Pandas operation:** `mean()`, `max()`, and `median()` calculate the average, highest value, and middle value of the Marks column.

### Q4. Group-wise Analysis

**Question 1:** Predict the department-wise average salary.

**Question 2:** Explain what the average salary tells us about each department.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Department': ['IT', 'HR', 'IT', 'HR'],
    'Salary': [40000, 30000, 60000, 50000]
})

print(df.groupby('Department')['Salary'].mean())

### Answer

**1. Predicted Output:**

```text
Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64
```

- **HR Average Salary:** 40000.0
- **IT Average Salary:** 50000.0

**2. Explanation:**  
The average salary is 40,000 for HR and 50,000 for IT.

**Pandas operation:** `groupby()` groups the data by Department, and `mean()` calculates the average salary for each department.

### Q5. Positive Correlation

**Question 1:** Predict the correlation coefficient between Hours and Marks.

**Question 2:** Interpret the direction and strength of the correlation.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))

### Answer

**1. Predicted Output:**

```text
1.0

**2. Explanation:**  
The relationship is positive and perfect. This means Hours and Marks increase together.

**Pandas operation:** `corr()` calculates the correlation coefficient between the Hours and Marks columns.

### Q6. Negative Correlation

**Question 1:** Predict the correlation coefficient between Price and Demand.

**Question 2:** Explain what the sign of the coefficient indicates.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Price': [10, 20, 30, 40],
    'Demand': [100, 80, 60, 40]
})

print(df['Price'].corr(df['Demand']))

### Answer

**1. Predicted Output:**

```text
-1.0

**2. Explanation:**  
The negative sign indicates that Price and Demand move in opposite directions. The value -1.0 indicates a perfect negative linear relationship.

**Pandas operation:** `corr()` calculates the correlation coefficient between the Price and Demand columns.

## Section B — Code Correction

### Q1. JSON/CSV

**Question:** A CSV file contains a header row, but the following code incorrectly reads the first student's record as column names. Correct the code.

In [ ]:
import pandas as pd

df = pd.read_csv("students.csv", header=None)

print(df)

Error: The header=None argument is incorrect because the CSV file already contains a header row. It causes the first row of the CSV file to be treated as data instead of column names.

In [ ]:
import pandas as pd

df = pd.read_csv("students.csv")

print(df)

### Answer

**1. Issue:**  
`header=None` tells Pandas that the CSV file has no header, so the first row is treated as data. This causes the first student's record to be used as column names.

**2. Correction:**  
Remove `header=None` because the CSV file already contains a header row.

**Corrected code:**

```python
df = pd.read_csv("students.csv")

Explanation:
By default, pd.read_csv() uses the first row of the CSV file as the column names.

### Q2. Data Analysis

**Question:** The objective is to display the five highest salaries from a DataFrame. Correct the code.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary').head(5))

Error: The code sorts salaries in ascending order, so it displays the lowest salaries first.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary', ascending=False).head(5))

### Answer

**1. Issue:**  
The code sorts the Salary column in ascending order, so the lowest salaries appear first.

**2. Correction:**  
Use `ascending=False` to sort the salaries from highest to lowest.

**Corrected code:**

```python
print(df.sort_values('Salary', ascending=False).head(5))

Explanation:
sort_values() sorts the salaries in descending order, and head(5) displays the five highest salaries.

Pandas operation: sort_values() sorts the Salary column, and head(5) selects the first five rows after sorting.

### Q3. Cleaning Data

**Question:** The following code fills the missing value with the mean, but the change is not stored back in the DataFrame. Correct the code.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Marks': [70, None, 90]
})

df['Marks'].fillna(df['Marks'].mean())

print(df)

**Error:** `fillna()` is not assigned back to `df['Marks']`, so the missing value is not updated in the DataFrame.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Marks': [70, None, 90]
})

df['Marks'] = df['Marks'].fillna(df['Marks'].mean())

print(df)

### Answer

**1. Issue:**  
The result of `fillna()` is not assigned back to the `Marks` column.

**2. Correction:**  
Assign the result back using `df['Marks'] =`.

**Corrected code:**

```python
df['Marks'] = df['Marks'].fillna(df['Marks'].mean())

### Q4. Correlation

**Question:** The objective is to calculate the correlation between Hours and Marks. Correct the code.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df.corr('Hours', 'Marks'))

**Error:** The `corr()` function is used incorrectly because it cannot take `Hours` and `Marks` as separate arguments.

In [ ]:
import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))

### Answer

**1. Issue:**  
`df.corr('Hours', 'Marks')` is incorrect because `corr()` cannot take two column names as separate arguments.

**2. Correction:**  
Select the `Hours` and `Marks` columns and use `corr()` between them.

**Corrected code:**

```python
print(df['Hours'].corr(df['Marks']))

## Section C — Edge Cases

### Q1. Broadcasting

**Question:** Two NumPy arrays have shapes (3, 1) and (1, 3). Predict the shape of their element-wise sum and explain how broadcasting works. Why might the result differ from adding two one-dimensional arrays?

**Given Code:**

```python
import numpy as np

a = np.array([[10], [20], [30]])
b = np.array([[1, 2, 3]])

result = a + b

print(result)
print(result.shape)
```

**Tasks:**

1. Predict the output of the given code.
2. Identify the shapes of arrays `a` and `b` before addition.
3. Explain how NumPy broadcasting enables the addition of these arrays.
4. What would happen if `b` were replaced with `np.array([1, 2])`? Explain the reason.
```

### Answer

**1. Predicted Output:**

```text
[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)
```

**2. Shapes of the arrays:**

- Array `a` has shape **(3, 1)**.
- Array `b` has shape **(1, 3)**.
- The resulting array has shape **(3, 3)**.

**3. Explanation of Broadcasting:**  
NumPy broadcasting allows the smaller dimensions of the arrays to expand so that the arrays can be added element-wise. Here, `(3, 1)` and `(1, 3)` are broadcast to `(3, 3)`.

**4. If b = np.array([1, 2]):**, 
NumPy broadcasting allows the arrays with shapes (3, 1) and (2,) to be added.
The result is:

[[11 12]
 [21 22]
 [31 32]]

The resulting shape is (3, 2).

**NumPy operation:** Broadcasting allows arrays with compatible shapes to perform element-wise operations without manually changing their shapes.

### Q2. CSV Edge Cases

**Question:** A CSV file contains commas inside quoted text fields, missing values, and dates in different formats. Explain how incorrect parsing can affect the resulting DataFrame and identify suitable read options or preprocessing steps.

**Given dataset:**

```text
students.csv

ID,Name,Date,Marks
101,"Asha, R",2026-09-01,80
102,Ravi,01/09/2026,
103,Neha,2026-09-03,NA
```

**Given code:**

```python
import pandas as pd

df = pd.read_csv("students.csv")

print(df)
print(df.dtypes)
```

**Tasks:**

1. Predict how the quoted comma in `"Asha, R"` will be handled when the CSV file is read.
2. Identify the potential issues with the Date and Marks columns.
3. Rewrite the code to handle missing marks and convert the date column into a suitable date-time format.
4. Explain why the value `"NA"` should be investigated before deciding whether it represents a missing value.

### Answer

**1. Quoted comma:**  
The comma inside `"Asha, R"` is treated as part of the Name because the value is enclosed in quotation marks. It will not create an extra column.

**2. Potential issues:**
- The `Date` column contains different date formats, such as `2026-09-01` and `01/09/2026`.
- The `Marks` column contains a missing value and `NA`.

**3. Corrected code:**

```python
df['Marks'] = pd.to_numeric(df['Marks'], errors='coerce')
df['Date'] = pd.to_datetime(df['Date'], errors='coerce', dayfirst=True)

print(df)
print(df.dtypes)
```

**4. Explanation:**  
The value `NA` should be investigated because it may represent a missing mark, or it may have another meaning in the original data. We should understand its meaning before treating it as a missing value.

**Pandas operation:** `pd.to_numeric()` handles numeric values and missing values, while `pd.to_datetime()` converts the Date column into a date-time format.

### Q3. Salary Outlier

**Question:** A salary column contains the values 25000, 30000, 35000, and 10000000. Explain how the extreme value affects the mean compared with the median. Identify which statistic may better represent a typical employee's salary and justify your choice.

**Given Code:**

```python
import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})

print("Mean:", df["Salary"].mean())
print("Median:", df["Salary"].median())
```

**Tasks:**

1. Predict the mean and median salary.
2. Identify the effect of the extreme salary on the mean.
3. Which statistic may better represent a typical employee's salary in this dataset? Justify your answer.
4. Write code to identify the employee record with the highest salary and explain why the record should be investigated before removing it.

### Answer

**1. Predicted Output:**

```text
Mean: 2522500.0
Median: 32500.0
```

**2. Effect of the extreme salary:**  
The salary `10000000` is much higher than the other salaries, so it increases the mean significantly. The median is much less affected by this extreme value.

**3. Statistic that better represents a typical salary:**  
The **median** may better represent a typical employee's salary because it is less affected by the extreme salary.

**4. Employee record with the highest salary:**

```python
highest_salary = df.loc[df["Salary"].idxmax()]

print(highest_salary)
```

**Output:**

```text
Employee           D
Salary      10000000
Name: 3, dtype: object
```

The record should be investigated before removing it because the high salary may be a genuine value or it may be a data-entry error. Removing it without checking could result in loss of valid information.

## Section D — Scenario-based Questions

### Q. Student Assessment Data Cleaning

**Question:** A student assessment dataset contains missing marks, duplicate student records, inconsistent department names such as IT, it, and I.T., and marks stored as text. Develop a cleaning sequence to standardize the data, handle missing values, remove duplicates, and convert marks into a suitable numeric type. State one risk of deleting missing values without investigating why they are missing.

**Using Pandas, perform the following cleaning operations in sequence:**

### 1. Load the CSV file and inspect the dataset using `head()`, `info()`, and `isnull().sum()`.

In [2]:
import pandas as pd

df = pd.read_csv("student_assessment_dirty.csv")

print(df.head())
print(df.info())
print(df.isnull().sum())

  Student_ID Student_Name Department Assessment   Marks
0       S101         Asha         IT     Python      85
1       S102         Ravi         it     Python      78
2       S103         Neha       I.T.     Python      92
3       S104        Karan         HR     Python     NaN
4       S105        Meera         HR     Python  Absent
<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Student_ID    15 non-null     str  
 1   Student_Name  15 non-null     str  
 2   Department    15 non-null     str  
 3   Assessment    15 non-null     str  
 4   Marks         12 non-null     str  
dtypes: str(5)
memory usage: 732.0 bytes
None
Student_ID      0
Student_Name    0
Department      0
Assessment      0
Marks           3
dtype: int64


### Explanation

`pd.read_csv()` loads the CSV file into a Pandas DataFrame. `head()` displays the first rows, `info()` shows the structure and data types, and `isnull().sum()` shows the number of missing values in each column.

### 2. Standardize department names so that IT, it, and I.T. are represented consistently. Apply the same approach to Finance.

In [3]:
df["Department"] = df["Department"].replace({
    "it": "IT",
    "I.T.": "IT",
    "finance": "Finance"
})

print(df["Department"])

0          IT
1          IT
2          IT
3          HR
4          HR
5     Finance
6     Finance
7          IT
8          IT
9          IT
10         HR
11    Finance
12         IT
13         HR
14    Finance
Name: Department, dtype: str


### Explanation

The department names are standardized so that `IT`, `it`, and `I.T.` are represented as `IT`. Similarly, `finance` is changed to `Finance`. This makes the department values consistent for analysis.

### 3. Identify and remove exact duplicate student assessment records.

In [4]:
df = df.drop_duplicates()

print(df)

   Student_ID Student_Name Department Assessment     Marks
0        S101         Asha         IT     Python        85
1        S102         Ravi         IT     Python        78
2        S103         Neha         IT     Python        92
3        S104        Karan         HR     Python       NaN
4        S105        Meera         HR     Python    Absent
5        S106        Arjun    Finance     Python         0
6        S107        Priya    Finance     Python        67
7        S108        Rahul         IT     Python        75
9        S109        Sneha         IT     Python       NaN
10       S110        Varun         HR     Python  88 marks
11       S111         Diya    Finance     Python        91
12       S112         Aman         IT     Python       NaN
13       S113        Nisha         HR     Python        64
14       S114        Kabir    Finance     Python       105


### Explanation

`drop_duplicates()` removes exact duplicate rows from the DataFrame. In this dataset, the duplicate record for student `S102` is removed.

### 4. Convert the Marks column into a numeric format, handling `"88 marks"`, `"Absent"`, `"NA"`, and blank entries appropriately.

In [5]:
df["Marks"] = pd.to_numeric(
    df["Marks"].astype(str).str.replace(" marks", "", regex=False),
    errors="coerce"
)

print(df["Marks"])

0      85.0
1      78.0
2      92.0
3       NaN
4       NaN
5       0.0
6      67.0
7      75.0
9       NaN
10     88.0
11     91.0
12      NaN
13     64.0
14    105.0
Name: Marks, dtype: float64


### Explanation

The Marks column is converted into a numeric format. The text `88 marks` is converted to `88`, while `Absent`, `NA`, and blank values are treated as missing values.

### 5. Investigate missing marks and distinguish absent students from students whose marks have not been entered. State one risk of deleting missing values without investigating why they are missing.

In [6]:
print(df[df["Marks"].isnull()][["Student_ID", "Student_Name", "Marks"]])

   Student_ID Student_Name  Marks
3        S104        Karan    NaN
4        S105        Meera    NaN
9        S109        Sneha    NaN
12       S112         Aman    NaN


### Explanation

The missing marks should be investigated because `Absent` may mean the student did not attend the assessment, while blank or `NA` may mean the marks were not entered.

**Risk of deleting missing values:**  
Deleting missing values without investigating may remove valid information and may incorrectly remove students who were absent.

### 6. Identify marks outside the valid range of 0–100 and decide how they should be handled.

In [7]:
invalid_marks = df[(df["Marks"] < 0) | (df["Marks"] > 100)]

print(invalid_marks)

   Student_ID Student_Name Department Assessment  Marks
14       S114        Kabir    Finance     Python  105.0


### Explanation

Marks should be within the valid range of 0 to 100. The value `105` is outside this range, so it is identified as an invalid mark.

The invalid value should be investigated before deciding whether to correct or remove it.

### 7. Display the cleaned DataFrame and report the number of records removed, missing marks remaining, and invalid marks identified.

In [8]:
print(df)

print("Records removed:", 1)
print("Missing marks remaining:", df["Marks"].isnull().sum())
print("Invalid marks identified:", len(invalid_marks))

   Student_ID Student_Name Department Assessment  Marks
0        S101         Asha         IT     Python   85.0
1        S102         Ravi         IT     Python   78.0
2        S103         Neha         IT     Python   92.0
3        S104        Karan         HR     Python    NaN
4        S105        Meera         HR     Python    NaN
5        S106        Arjun    Finance     Python    0.0
6        S107        Priya    Finance     Python   67.0
7        S108        Rahul         IT     Python   75.0
9        S109        Sneha         IT     Python    NaN
10       S110        Varun         HR     Python   88.0
11       S111         Diya    Finance     Python   91.0
12       S112         Aman         IT     Python    NaN
13       S113        Nisha         HR     Python   64.0
14       S114        Kabir    Finance     Python  105.0
Records removed: 1
Missing marks remaining: 4
Invalid marks identified: 1


### Explanation

The cleaned DataFrame is displayed after removing the duplicate record and standardizing the department names.

- **Records removed:** 1 duplicate record
- **Missing marks remaining:** Marks that are blank, `Absent`, or `NA` remain missing.
- **Invalid marks identified:** `105` is outside the valid range of 0–100 and should be investigated.

The final report summarizes the cleaning results.